README.md

```diff
@@ -104,7 +104,98 @@ module.exports = {
 
 Note that while the above example tests a webpage with [puppeteer](https://github.com/GoogleChrome/puppeteer) and uses pentf's has native support for HTTP requests (in `net_utils`) and email sending (in `email`), tests can be anything – they just have to fail the promise if the test fails.
 
-Have a look in the [API documentation](https://boxine.github.io/pentf/index.html) for various helper functions.
+Have a look in the [API documentation](https://boxine.github.io/pentf/) for various helper functions.
+
+## Tips for writing good tests
+
+By their very nature, end-to-end tests can be flaky, i.e. sometimes succeed and sometimes fail when run multiple times. We want the tests to only relay the flakiness of the systems we test, and not introduce any additional flakiness ourselves. Here are a few tips for that:
+
+### Use `data-testid` attributes in browser tests
+
+Unless otherwise documented, class names and document structures are subject to change. By setting an explicit attribute like `data-testid="comment-button"` in the code, the developer and tester set up a stable contract.
+
+**Avoid**: `await page.waitForSelector('h1 > div.main-container form p button.large-button');`
+
+**Use**: `await waitForTestId(page, 'comment-button');`
+
+### Always wait a bit
+
+If the tests run quick and a remote system is slow, the browser UI may not update immediately. If there is high local system load (which is necessary for the tests to run quickly), then even local updates may not be immediate:
+
+**Avoid**: `page.$('foo')`, `page.evaluate(() => document.querySelector('foo'))`
+
+**Use**: `page.waitForSelector('foo')`
+
+### Make sure you start waiting _early_
+
+Make sure that an action you are waiting on has not already happend. In particular, `await page.waitForNavigation()` calls should probably be replaced by checks for the new page.
+
+**Avoid**:
+```javascript
+await page.press('Enter');
+// If the page is quick, navigation may have already occured here!
+await page.waitForNavigation();
+await waitForText(page, 'email sent');
+
+const since = new Date(); // Too late, email may already have been sent!
+await getMail(config, since, email, 'Enter was pressed');
+```
+
+**Use**:
+```javascript
+const since = new Date(); // Too late, email may already have been sent!
```

Comment:
The comment shouldn't be here anymore, right?

Reply:
You're totally right. See #69 for fix.